File: moqt/internal/message/subscribe_ok.go
package message

import (
	"io"
)

/*
 * SUBSCRIBE_OK Message {
 *   Group Frequency (varint),
 * }
 */
type SubscribeOkMessage struct {
}

func (som SubscribeOkMessage) Len() int {
	return 0
}

func (som SubscribeOkMessage) Encode(w io.Writer) error {
	msgLen := som.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _ = WriteMessageLength(b, uint64(msgLen))

	_, err := w.Write(b)

	return err
}

func (som *SubscribeOkMessage) Decode(src io.Reader) error {
	num, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	b := make([]byte, num)
	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: interop/web/src/message/subscribe_ok.ts
import { Writer, Reader } from "../io";
import { varintLen } from "../io/len";

export class SubscribeOkMessage {
    groupOrder: bigint;

    constructor(groupOrder: bigint) {
        this.groupOrder = groupOrder;
    }

    length(): number {
        return varintLen(this.groupOrder);
    }

    static async encode(writer: Writer, groupOrder: bigint): Promise<[SubscribeOkMessage?, Error?]> {
        const msg = new SubscribeOkMessage(groupOrder);
        writer.writeVarint(BigInt(msg.length()));
        writer.writeVarint(groupOrder);
        const err = await writer.flush();
        if (err) {
            return [undefined, err];
        }
        return [msg, undefined];
    }

    static async decode(reader: Reader): Promise<[SubscribeOkMessage?, Error?]> {
        const [len, err] = await reader.readVarint();
        if (err) {
            return [undefined, new Error("Failed to read length for SubscribeOkMessage: " + err.message)];
        }

        const [varint, err2] = await reader.readVarint();
        if (err2) {
            return [undefined, new Error("Failed to read groupOrder for SubscribeOkMessage: " + err2.message)];
        }

        return [new SubscribeOkMessage(varint), undefined];
    }
}